### PR TITLE
fix batchnorm infer dtype failed in half inference

### PR DESCRIPTION
**Pull request by BBuf**
#close https://github.com/Oneflow-Inc/oneflow/issues/9381

### CODE REVIEW

File: oneflow/core/functional/impl/nn_functor.cpp [hunk 1]
```diff
@@ -2062,17 +2062,35 @@ class NormalizationFunctor {
       beta_val = JUST(functional::Constant(gamma_beta_shape, 0.0, x->dtype(), JUST(x->device())));
     }
 
+    if (gamma_val->dtype()->data_type() == DataType::kFloat16 || gamma_val->dtype()->data_type() == DataType::kBFloat16) {
```
Comment (BBuf) [inline oneflow/core/functional/impl/nn_functor.cpp:2065]: 在normalization op的type推导时限制了moving_mean, moving_var，gamma，beta不能为fp16，所以这里使用一个cast将model.half()之后的各项转换成fp32。

File: python/oneflow/test/modules/test_batchnorm.py [hunk 2]
```diff
@@ -80,6 +80,26 @@ def test_batchnorm3d_module_with_random_data(test_case):
...
+            track_running_stats=random().to(bool),
+            affine=random().to(bool),
+        ).to(device)
+        m.train(random())
+        m.half()
+        x = random_tensor(
+            ndim=4, dim0=random(1, 4), dim1=channel, requires_grad=True
+        ).to(device)
+        x.half()
+        y = m(x)
```
Comment (wyg1997) [inline python/oneflow/test/modules/test_batchnorm.py:100]: 是不是可以考虑跑 2 次，判断 2 次的结果不同？

**Review by wyg1997 — APPROVED** [review]

File: python/oneflow/test/modules/test_batchnorm.py (hunk 2 above)
Comment (BBuf) [reply python/oneflow/test/modules/test_batchnorm.py:100]: 自动测试这里不太好写，我感觉这个测试已经可以说明moving_mean和moving_var已经被原地更新了。

Comment (BBuf) [reply python/oneflow/test/modules/test_batchnorm.py:100]: 我试过如果不原地更新，最后的moving_mean和moving_var是过不了精度检查的

Comment (wyg1997) [reply python/oneflow/test/modules/test_batchnorm.py:100]: 哦哦也是，那没问题了

**Review by Flowingsun007 — APPROVED** [review]
